controller/kv/helper.go

```diff
@@ -1582,6 +1627,227 @@ func (m clusterHelper) GetScannerDB(store string) []*share.CLUSScannerDB {
+			log.WithFields(log.Fields{"error": err, "scanner": selectedScanner.ID}).Error("Failed to track credit acquisition")
+
+			// Release the credit back to the scanner
+			releaseCreditErr := m.ReleaseScanCredit(selectedScanner.ID, 1)
+			if releaseCreditErr != nil {
+				log.WithFields(log.Fields{"error": releaseCreditErr, "scanner": selectedScanner.ID}).Error("Failed to release scanner credit")
+			}
+			return nil, releaseCreditErr
+		}
+
+		return selectedScanner, nil
+	}
+
+	return nil, share.ErrNoScannerAvailable
+}
+
+func (m clusterHelper) ReleaseScanCredit(scannerId string, releaseCredit int) error {
+	key := share.CLUSScannerKey(scannerId)
+	for retry := 0; retry < m.maxClusterOperationRetries; retry++ {
+		// Get current scanner state with revision
+		scanner, rev, err := m.GetScannerRev(scannerId)
+		if err != nil {
+			// Check if scanner was deleted (object not found)
+			if errors.Is(err, common.ErrObjectNotFound) {
+				// Scanner has been deleted - credit is implicitly released with the scanner
+				log.WithFields(log.Fields{"scanner": scannerId}).Debug("Scanner not found during credit release, likely deleted")
+				return nil
+			}
+
+			// Network error or other temporary failure - retry with backoff
+			log.WithFields(log.Fields{"error": err, "scanner": scannerId}).Warn("Failed to get scanner, retrying")
+			continue
+		}
+
+		// Increment credit, ensuring it doesn't exceed max
+		scanner.ScanCredit = min(scanner.ScanCredit+releaseCredit, scanner.MaxConcurrentScansPerScanner)
+
+		// Attempt to update using CAS (PutRev)
+		value, err := json.Marshal(scanner)
+		if err != nil {
+			log.WithFields(log.Fields{"error": err, "scanner": scannerId}).Error("Failed to marshal scanner")
+			return err
+		}
+
+		// Use PutRev for atomic compare-and-swap
+		if err := cluster.PutRev(key, value, rev); err != nil {
+			if errors.Is(err, cluster.ErrPutCAS) {
+				// CAS failed - either revision mismatch or scanner was deleted
+				// Both cases are handled by retrying (scanner deletion will be caught in next iteration)
+				log.WithFields(log.Fields{"error": err, "scanner": scannerId}).Warn("Failed to release scanner credit, retrying")
+				continue
+			}
+			return err
+		}
+
+		if err := m.TrackCreditAcquisition(m.id, scannerId, scanner.MaxConcurrentScansPerScanner, false); err != nil {
+			log.WithFields(log.Fields{"error": err, "scanner": scannerId}).Error("Failed to track credit acquisition")
+			return err
+		}
+
+		return nil
+	}
+
+	return fmt.Errorf("failed to release scanner credit after %d retries", m.maxClusterOperationRetries)
+}
+
+func (m clusterHelper) initCreditOwners() error {
+	key := share.CLUSScannerCreditOwnerKey(m.id)
+	scanners := make(map[string]int)
+
+	var value []byte
+	var err error
+	value, err = json.Marshal(scanners)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < m.maxClusterOperationRetries; i++ {
+		if err = cluster.PutRev(key, value, 0); err == cluster.ErrPutCAS {
+			continue
+		}
+		break
+	}
```

Comment:
We probably don't need to call `json.Marshal()` since it's always empty?

```suggestion
	scanners := make(map[string]int)
	var err error

	for i := 0; i < m.maxClusterOperationRetries; i++ {
		if err = cluster.PutRev(key, []byte("{}"), 0); err == cluster.ErrPutCAS {
			continue
		}
		break
	}
```